django/db/models/functions/datetime.py

```diff
@@ -12,31 +12,60 @@
 from django.utils.functional import cached_property
 
 
-class Extract(Transform):
+class TimezoneMixin(object):
+    tzinfo = None
+
+    def get_tzname(self):
+        # Timezone conversions must happen to the input datetime *before*
+        # applying a function. 2015-12-31 23:00:00 -02:00 is stored in the
+        # database as 2016-01-01 01:00:00 +00:00. Any results should be
+        # based on the input datetime not the stored datetime.
+        tzname = None
+        if settings.USE_TZ:
+            if self.tzinfo is None:
+                tzname = timezone.get_current_timezone_name()
+            else:
+                tzname = timezone._get_timezone_name(self.tzinfo)
+        return tzname
+
+
+class Extract(Transform, TimezoneMixin):
     lookup_name = None
 
-    def __init__(self, expression, lookup_name=None, **extra):
+    def __init__(self, expression, lookup_name=None, tzinfo=None, **extra):
         if self.lookup_name is None:
             self.lookup_name = lookup_name
         if self.lookup_name is None:
             raise ValueError('lookup_name must be provided')
+        self.tzinfo = tzinfo
         super(Extract, self).__init__(expression, **extra)
 
     def as_sql(self, compiler, connection):
         sql, params = compiler.compile(self.lhs)
         lhs_output_field = self.lhs.output_field
         if isinstance(lhs_output_field, DateTimeField):
-            tzname = timezone.get_current_timezone_name() if settings.USE_TZ else None
+            tzname = self.get_tzname()
             sql, tz_params = connection.ops.datetime_extract_sql(self.lookup_name, sql, tzname)
             params.extend(tz_params)
         elif isinstance(lhs_output_field, DateField):
             sql = connection.ops.date_extract_sql(self.lookup_name, sql)
         elif isinstance(lhs_output_field, TimeField):
             sql = connection.ops.time_extract_sql(self.lookup_name, sql)
         else:
-            raise ValueError('Extract only valid on Date/Time/DateTimeFields')
+            assert False, "Tried to Extract from an invalid type."
```

Comment:
This line should never be hit because resolve_expression below already validates the output_field. I'd prefer it to be an assertion for that reason.